libp2p/protocols/autonat.nim

```diff
@@ -0,0 +1,276 @@
+                         addrs: addrs
+                       ))).encode()
+  await conn.writeLp(pb.buffer)
+
+proc sendResponseError(conn: Connection, status: ResponseStatus, text: string = "") {.async.} =
+  let pb = AutonatDialResponse(
+             status: status,
+             text: if text == "": none(string) else: some(text),
+             ma: none(MultiAddress)
+           ).encode()
+  await conn.writeLp(pb.buffer)
+
+proc sendResponseOk(conn: Connection, ma: MultiAddress) {.async.} =
+  let pb = AutonatDialResponse(
+             status: ResponseStatus.Ok,
+             text: some("Ok"),
+             ma: some(ma)
+           ).encode()
+  await conn.writeLp(pb.buffer)
+
+type
+  Autonat* = ref object of LPProtocol
+    switch*: Switch
+
+proc dialBack*(a: Autonat, pid: PeerId, ma: MultiAddress|seq[MultiAddress]):
+    Future[MultiAddress] {.async.} =
+  let addrs = when ma is MultiAddress: @[ma] else: ma
+  let conn = await a.switch.dial(pid, addrs, AutonatCodec)
+  defer: await conn.close()
+  await conn.sendDial(a.switch.peerInfo.peerId, a.switch.peerInfo.addrs)
+  let msgOpt = AutonatMsg.decode(await conn.readLp(1024))
+  if msgOpt.isNone() or
+     msgOpt.get().msgType != DialResponse or
+     msgOpt.get().response.isNone():
+    raise newException(AutonatError, "Unexpected response")
+  let response = msgOpt.get().response.get()
+  if response.status != ResponseStatus.Ok:
+    raise newException(AutonatError, "Bad status " &
+                                      $response.status & " " &
+                                      response.text.get(""))
+  if response.ma.isNone():
+    raise newException(AutonatError, "Missing address")
+  return response.ma.get()
+
+proc doDial(a: Autonat, conn: Connection, addrs: seq[MultiAddress]) {.async.} =
+  try:
+    let ma = await Dialer(a.switch.dialer).canDial(conn.peerId, addrs, @[AutonatCodec])
+    await conn.sendResponseOk(ma)
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    await conn.sendResponseError(DialError, exc.msg)
+
+proc handleDial(a: Autonat, conn: Connection, msg: AutonatMsg): Future[void] =
+  if msg.dial.isNone() or msg.dial.get().peerInfo.isNone():
+    return conn.sendResponseError(BadRequest, "Missing Peer Info")
+  let peerInfo = msg.dial.get().peerInfo.get()
+  if peerInfo.id.isSome() and peerInfo.id.get() != conn.peerId:
+    return conn.sendResponseError(BadRequest, "PeerId mismatch")
+
+  if conn.observedAddr[0].isErr() or (not IP4.match(conn.observedAddr[0].get()) and
+                                      not IP6.match(conn.observedAddr[0].get())):
+    return conn.sendResponseError(InternalError, "Expected an IP address")
+
+  var addrs = initHashSet[MultiAddress]()
+  addrs.incl(conn.observedAddr)
+  for ma in peerInfo.addrs:
+    let maFirst = ma[0]
+    if maFirst.isErr() or (not IP4.match(maFirst.get()) and
+                           not IP6.match(maFirst.get())):
+      continue
+    addrs.incl(ma)
```

Comment:
If I  understand it correctly, you are currently dialing any address that is sent in `PeerInfo.addrs` (up to the `ArbitraryLimit`). Note that for DoS protection the server must only dial addresses that are based on `observedAddr`. Quoting from the AutoNAT specs: 
> In order to prevent attacks like the one described in [RFC 3489, Section 12.1.1](https://www.rfc-editor.org/rfc/rfc3489#section-12.1.1) (..), implementations MUST NOT dial any multiaddress unless it is based on the IP address the requesting node is observed as. This restriction as well implies that implementations MUST NOT accept dial requests via relayed connections as one can not validate the IP address of the requesting node.

In go-libp2p and rust-libp2p we solve this by replacing each addresses' IP (in your case `maFirst`) with the IP of `observedAddr`:

- [1] Logic in Go: https://github.com/libp2p/go-libp2p/blob/3d351e4ed396917866f464f2ab90c838fe7d92a2/p2p/host/autonat/svc.go#L154-L168
- [2] Logic in Rust:  https://github.com/libp2p/rust-libp2p/blob/d0da3a0973431babba94fffd3ffd7723e0b09837/protocols/autonat/src/behaviour/as_server.rs#L313-L352

Reply:
@elenaf9 @mxinden should we update the spec to reflect the fact that wrong addresses will be "fixed" using the observed address?

The current spec really reads as "any address not matching the observed address will be __skipped__"

Reply:
I am in favor of such change @Menduist. Would you mind opening a pull request?